### DISCUSSION TITLE
Shared state between browser and server

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#4095] [Ideas / Feature Requests]
### Question

As title says, I'm looking for better ways to communicate between browser scripts and nicegui backend.

The usual way is `ui.run_javascript`.  But it's a very blunt instrument.  Among other issues:
- js code saved as python string, so no syntax checking or highlighting
- requires knowing js code to do what you need
- error messages end up on browser console
- requires `await` on results, so caller must be `async`

Does anyone have a more elegant way to communicate?

# Desired 

Ideally what would be better is a standard interface for nicegui and browser scripts to exchange data.  Nicegui already does this for page components: browser loads a web worker or such, web worker opens socket to nicegui server, two sides exchange data to keep page DOM and server in sync.  I haven't looked into details, but generally it's something like that.

I imagine the nicegui / browser script interface could work as follows:
- web worker defines a location for shared data, such as `window.ngshared` which can be a dict / object
- browser script puts data into `window.ngshared` such as `window.ngshared ["foo"] = "bar"`
- web worker detects change to `window.ngshared` and mirrors to nicegui server
- nicegui page (server code) reads from `window.ngshared ["foo"]` to get current value of "foo"
- and vice versa for sending data from server to browser script

For simplicity, data can be limited to json serializable objects like strings, lists, dicts, etc.

**Comment by ed2050** [comment]
more background info...

# Purpose

Why is this useful?  For one, it avoids the issues above with `ui.run_javascript`.  Current data is always available.  Browser scripts can be more complex than run_js snippets and live in its own js file with syntax highlighting, typechecking, etc.  It also enables using any language for browser scripts, not just js: python (brython), typescript, f-sharp, anything that compiles to web assembly, and more.

It's also useful because nicegui server doesn't and can't track all browser state.  Example: what items are current visible on the page, i.e. scrolled into view?  Server doesn't know, only the browser tracks that.

# Examples

I use browser scripts for client-side functionality that doesn't work well in nicegui.  Say you have a page with many images arranged vertically.  You want Pageup and Pagedown keys to scroll exactly one image.  Can you do it in nicegui?  Sure, you could implement a keyboard listener, keep track of current image at top of page, find next / previous image on keypress, and call `ui.run_javascript ("doc.getElement (nextimg).scrollIntoView()")`.  Or you could take care of that in a browser script, saving laggy round trip communication just to scroll the page.

Now imagine user scrolls the page freely with mouse wheel.  They see an image they want and press a key to copy it.  How do you find which image is currently in view?
You could:
- have nicegui server code call `ui.run_js (img.getBoundingClientRect ())` for every image on page, wait for results, and check positions
- have a browser script function `get_top_image` which runs all checks in browser and just returns the relevant image id.  call that function from nicegui with `ui.run_js (get_top_image())`.  current best solution.
- have a shared data area `window.ngshared`.  browser script listens for scroll events and updates `window.ngshared ["topimg"]` accordingly.  when nicegui keyboard listener sees 'c' keypress, all it has to do is read current `nshared ["topimg"]` value and it immediately has the correct image.

Those are just a few examples.  I have other situations where behavior is implemented client-side because doing it in nicegui is difficult, inefficient, error prone, or just impossible.  Imagine implementing a floating magnifier box like on Amazon.  Things get rather complex so I'll leave it here for now.

**Comment by ed2050** [comment]
# Possible Solution

Here's one possible solution.  Not a great one, but better than `run_javascript`.  Any feedback?

Add two `input` elements to nicegui page.  Bind one for incoming communication (from browser script to ng server) and other for outgoing (from ng server to browser script).  Elements can be invisible with css `display : none`.

```py
# --- nicegui ---

@ui.page ('/foo')
def foopage () :
    pagedata = { 'toserver' : '' , 'toclient' : '' }
    ui.input ().props ('id=toclient').bind_value (pagedata, 'toclient')
    ui.input ().props ('id=toserver').bind_value (pagedata, 'toserver')

    # send message to browser script
    pagedata ['toclient'] = 'foo'

    # read message from browser script
    inmsg = pagedata ['toserver']

# --- browser script (brython) ---
def fooclient () :
    # nicegui / quasar puts id on <label> parent of <input>
    inbox = doc.querySelector ('#toclient input')
    outbox = doc.querySelector ('#toserver input')
    
    # send message to nicegui server
    outbox.value = 'oof'

    # read message from nicegui server
    msg = inbox.value
```

If nicegui webworkers do their job then changes made on one end should be mirrored to the other.

I haven't tested.  There's a chance that setting `outbox.value` in client script won't be enough to trigger mirroring.  Depends whether client side component listens for DOM events _[input](https://developer.mozilla.org/en-US/docs/Web/API/Element/input_event)_ (not triggered by script changes) or _[change](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)_ or both.  Depending on what it listens for, client side send can be a bit more complicated: may need to create and dispatch a pseudo-event on the input component to trigger.

# Issues

It's far from a perfect solution.  Among the issues:

- Communication is text only.  This can be somewhat improved by serializing messages to json.

- Message size limited to max `input` length.  Can use `textarea` if more space needed.
 
- Limited to one message at a time.  If inbox value changes before message is processed, then message is lost.  This could be addressed via a message container format with message numbering and appending new messages instead of overwriting value.  But then `input.value` will grow unbounded.

- Hard to synchronize messages.  When server sends new message, how can it know whether client received and processed previous message?  And vice versa.  Using a message queue can avoid losing messages (see above), but still don't know when or if a message has been actioned.  Probably best used for simple data where "give me a recent value" is good enough.

You could go even more complicated, building multiple message pipes, message numbering, reply values, etc.  But that seems silly.  For simple messages with one reply, `run_js` is good enough for server to query client, and `@app.get` for client to query server.

This is the closest I could come to ideal shared-state solution with current nicegui code.  But I don't know ng internals.  Maybe someone has a better way to synchronize data between server and browser script?

**Comment by falkoschindler** [comment]
Hi @ed2050,

Let's address your issues with `ui.run_javascript`:

> - js code saved as python string, so no syntax checking or highlighting

True. But there are IDE extensions like https://marketplace.visualstudio.com/items?itemName=DaelonSuzuka.nicegui that support syntax highlighting in nested JavaScript strings.

> - requires knowing js code to do what you need

Yes, but for implementing your custom frontend code you need JavaScript anyway, don't you?

> - error messages end up on browser console

Sure. But this would also happen if there's an error while accessing your shared data storage. How else should we handle errors in the frontend? And what's the problem with using the JavaScript console for debugging frontend code?

> - requires await on results, so caller must be async

If you want to run JavaScript code that returns a response, of course you need to await it. But if that's not a requirement, you can use other way to send data back to the server, like [generic events](https://nicegui.io/documentation/generic_events#custom_events).

For more complex solutions, [custom components](https://nicegui.io/documentation/section_configuration_deployment#custom_vue_components) already provide what’s needed for ongoing data exchange between browser code and NiceGUI’s server. They let you integrate your own JS (or Brython, WASM, etc.) into the NiceGUI framework and connect to the built-in WebSocket synchronization mechanism. This eliminates manual calls to ui.run_javascript and allows you to define your own props, events, and updates in a clean, official way rather than relying on hidden inputs or other “hacks.”

**Reply by ed2050** [reply]
> For more complex solutions, [custom components](https://nicegui.io/documentation/section_configuration_deployment#custom_vue_components) already provide what’s needed for ongoing data exchange between browser code and NiceGUI’s server. They let you integrate your own JS (or Brython, WASM, etc.) into the NiceGUI framework and connect to the built-in WebSocket synchronization mechanism. This eliminates manual calls to ui.run_javascript and allows you to define your own props, events, and updates in a clean, official way rather than relying on hidden inputs or other “hacks.”

Thanks @falkoschindler!  This might potentially work.  Behavior sounds quite similar to what I want.  The problem is, I don't use vue and don't write js.  Let me look into it and see.

I considered custom components to be for special purpose widgets like a color picker.  Having one general-purpose component for mirroring shared values with client-side scripts seems like something that would be useful for many people.  Advanced users to be sure, but no more than other built-in advanced features.

To clarify a few other points...

> True. But there are IDE extensions like https://marketplace.visualstudio.com/items?itemName=DaelonSuzuka.nicegui that support syntax highlighting in nested JavaScript strings.

I don't use Visual Studio, I use vim.  And it's more than just syntax highlighting.  Embedding multiline js code with proper indentation makes a mess of your python.  And line numbers in error messages are useless for embedded js strings.  Complex client-side code belongs together in a separate file.

> Yes, but for implementing your custom frontend code you need JavaScript anyway, don't you?

No actually.  I run client scripts in [brython](https://brython.info/) aka browser python.  I only know a bit of js and avoid it whenever possible.  Terrible ugly language with too many gotchas.

> > * error messages end up on browser console
> 
> Sure. But this would also happen if there's an error while accessing your shared data storage. How else should we handle errors in the frontend? And what's the problem with using the JavaScript console for debugging frontend code?

I'm fine with console messages.  The issue is, I have no idea when there's an error.  nicegui doesn't register a problem when run_js call fails.  You'd have to put code in your `run_javascript` snippet to detect and flag errors, which gets complicated.

It's much easier when running a full client-side script.  Script has error checking, can throw up a message box or such on errors.  That gives you a general idea what happened ("Error, no image found") and you go to console for more info.

> > * requires await on results, so caller must be async
> 
> If you want to run JavaScript code that returns a response, of course you need to await it. But if that's not a requirement, you can use other way to send data back to the server, like [generic events](https://nicegui.io/documentation/generic_events#custom_events).

My point was that the async / non-async distinction is a mess.  I write a bunch of regular functions that all work great.  Later on, I improve my app by having one of the functions `def foo()` retrieve data that requires `run_javascript`.  Now `foo` needs to use `await`.  

But oops!  `foo` is not an async function so using await is an error.  Ok just add `async` to `def foo()` and we're good right?

Ooops!  Now every function `def bar ()` that calls `foo()` has to add `await foo()`.  

Ooops!  But `bar` isn't async, so now we need to add `async def bar ()`.  

Oops!  .... repeat ad inifinitum.

Async is a really crappy system.  The simple act of calling an async function propagates pointless changes through your entire codebase.  [See this for more](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

## Custom events 

How would custom events work for retrieving a current value?  Say browser script emits an event every time value changes.  So nicegui needs a custom listener just to watch for changes on this value?

Maybe that works ok for a couple values.  Doesn't scale well when you're watching dozens of values.  Doesn't help with sending value changes the other direction (from server to client).  Seems limited to a few specific circumstances.

Why not have a general bidirectional pipe that mirrors value changes from both ends?  That's exactly what nicegui does already for controls like `input`, `slider`, etc.  Why not add one more bidirectional dict / object for sharing state that isn't tied to a control widget?

**Reply by falkoschindler** [reply]
Ok, you made some valid points, which, however, are a little out of scope of this project (choice of IDE, nesting different languages in one file, colored languages, pro/contra JavaScript in the frontend, ...). Let's focus on your idea:

I'm starting to understand that a "shared object" would be handy. Maybe not a hidden `ui.input`, but maybe a hidden `ui.json_editor` - just without the UI element around it. Deriving from `ValueElement` would automatically raise change events and provide bindings.

So far my mindset was: Usually you don't have to write JavaScript, because everything is in Python. Maybe you need a little bit of JavaScript, then `ui.run_javascript` and `emitEvent` is fine for communication. And if you need even more control over the frontend, you can write a custom component. But maybe some kind of `ui.model` would be a useful tool.

**Reply by ed2050** [reply]
Yes, exactly.  It comes down to design philosophy.

The approach of "handle everything in server code" is a great starting point.  Nicegui does amazing things from a single codebase.  But eventually, you hit its limits.

You can keep piling on hacks that force everything into server-side model, like `run_javascript`.  Maintain ideological consistency, at the expense of practicality.  Or you can provide a first-class pipeline between server & client.  For needs that go beyond what server can do.

Ideological purity is already gone.  Nicegui has several mechanisms for client-side js.  `run_js` for setting doc title or scrolling to location on page.  Special `:classes` props with js [like here](https://github.com/zauberzeug/nicegui/discussions/4103).  Custom components.  These all expose client-side scripts to varying degrees.  A `ui.model` component would just create a data pipeline when snippets aren't enough.

`ui.json_editor` looks like a good place to start.  

# Page state

Perhaps another way to think about this shared object would be to call it "page state".  Many of my pages use a "page state" data class `ps` to hold data for the loaded page instance:
- ps.title for page title
- ps.request for request object
- ps.client for client object
- ps.filters for the collection of user-supplied filters in use (name, size, date, etc)

This keeps all data related to the current page localized, and easier to pass around - just send `ps` to every function, instead of passing `request` object explicitly to 3 nested functions that don't need it just for one function at the bottom to use it (and client, and title, and filters, ...).

Nicegui could provide a `pagestate` data object for manipulating certain page objects that currently require `run_js`.  Such as:
- `pagestate.title` could mirror the `<title>` tag in document `<head>`.  Then server could set `pagestate.title = "foo"` instead of `ui.run_javascript ('document.title = "foo"', respond = false)` (as [here](https://github.com/zauberzeug/nicegui/discussions/830)). 
- `pagestate.focused` (readonly) could mirror `document.hasfocus()` to tell whether page currently has focus.
- `pagestate.selected` (readonly) could mirror `document.getSelected ().toString()` for currently highlighted text.

Those would be default page items.  Then user code (server side or client side) could add its own keys and values, like `pagestate.foo = "bar"`.  With changes to keys and values mirrored on both server and client.  Like your proposed `ui.model` with a few default values common to all pages.

Just an idea.  Could eliminate some common uses of run_js while also providing client script communication.  Or maybe it's too much.  A purely user `ui.model` object works fine too.

**Comment by falkoschindler** [comment]
Ok, before digging deeper into the details of this feature request, let's hear the community. Would this be a valuable feature? Feel free to upvote the original post.
